Cython/Utility/ObjectHandling.c

```diff
@@ -309,7 +304,13 @@ static PyObject *__Pyx_PyObject_GetItem(PyObject *obj, PyObject* key) {
     if (likely(m && m->mp_subscript)) {
         return m->mp_subscript(obj, key);
     }
-    return __Pyx_PyObject_GetIndex(obj, key);
+    PySequenceMethods *sm = Py_TYPE(obj)->tp_as_sequence;
+    if (likely(sm && sm->sq_item)) {
+        return __Pyx_PyObject_GetIndex(obj, key);
+    }
+
+    // Fall back on CPython's GetItem for non-mappings, non-sequences.
+    return PyObject_GetItem(obj, key);
```

Comment:
I think the issue with this is that `PyObject_GetItem` ends up repeating everything you've just done, so therefore this path is unnecessarily slow.

The original issue [linked to the commit that added this feature in CPython](https://github.com/python/cpython/pull/4732/files#diff-fa8da9109fa9e26d43de374c663544fe) so it'd be easy enough to take the code from there.

Reply:
Yes it seems better to just duplicate the logic in CPython's GetItem to deal with this. I'll update to do this.

One thing I'm unsure about in general is why cython duplicates the implementations of these functions. Are there optimizations that cython is making that can't be implemented within CPython? Do I need to take any of this into account when adapting the implementation from CPython?

Reply:
I didn't write the original code so I'm not 100% sure sure. I think there's two potential speedups I can see:

1. No `NULL` check at the start of the function (which I think is helped because Python has a pretty good idea of what can be NULL or not)
2. Special-case for tuple in `__Pyx_GetItemInt_Fast`.

I'm not 100% sure what you need to account for though.

Reply:
Alright, thanks so much for the help.

Updated the PR to pull in the implementation from CPython 3.7. It also looks like the PR is going to be a bit more complicated than I first thought as it seems we need to add support for implicit classmethods (correct me if I'm wrong here).

I probably need to add some logic around the python version as well when deciding what should be considered an implicit classmethod for total correctness here.

Additionally it looks like there's going to be needed some fixes to the mro logic to be fully compliant with PEP 560. Haven't looked totally at what's going to be required but the included test from CPython still has 7 failing mro-related tests.